### PR TITLE
feat: MTLS support for the InfluxDB Python client

### LINKED ISSUE #508: MTLS support for the Python client
__Proposal:__
We should also support mutual TLS for the InfluxDB client. The implementation could be handled as a quick win and I can prepare the corresponding functionality inside a PR. The client uses `urllib3` as a library to establish a connection to the InfluxDB and the [MTLS support](https://urllib3.readthedocs.io/en/stable/advanced-usage.html#client-certificates) is already implemented inside it. Currently, there are two possible ways to implement the MTLS support inside the client:

1. Forward the custom Python SSL context to the urllib3 via the `**kwags` functionality (Personal favorited implementation, more adaptable)
2. Use the same functionality as the normal TLS implementation and forward the certificates to the `urllib3`

__Current behavior:__
Currently, it's possible to define e.g. a Python SSL context as `**kwags` but this context is not forwarded to all calls at the end.

__Desired behavior:__
I would expect that MTLS is also supported by the library.

__Use case:__
The use case is the possibility to establish an MTLS connection to the InfluxDB.

### PR DESCRIPTION
**Pull request by ZPascal**
Closes #508

## Proposed Changes

Add MTLS support to the InfluxDB Python client as already discussed inside the following [issue](https://github.com/influxdata/influxdb-client-python/issues/508).

## TODO
- [x] Check the context usage (sync case)
- [x] Add unit tests
- [x] Squash the commits together 

## Checklist

- [x] CHANGELOG.md updated
- [x] Rebased/mergeable
- [x] A test has been added if appropriate
- [x] `pytest tests` completes successfully
- [x] Commit messages are [conventional](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Sign [CLA](https://www.influxdata.com/legal/cla/) (if not already signed)

### CODE REVIEW

**Comment by ZPascal** [comment]
Hi @bednar, should I also prepare an integration test scenario, or is it for these optional parameters not necessary?

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ## 1.34.0 [unreleased]
-
+1. [#509](https://github.com/influxdata/influxdb-client-python/pull/509): MTLS support for the InfluxDB Python client
```
Comment (bednar) [inline CHANGELOG.md:2]: Please add `### Features` header:

```suggestion
### Features
1. [#509](https://github.com/influxdata/influxdb-client-python/pull/509): MTLS support for the InfluxDB Python client
```

File: README.rst [hunk 2]
```diff
@@ -197,6 +197,9 @@ The following options are supported:
 - ``timeout`` - socket timeout in ms (default value is 10000)
 - ``verify_ssl`` - set this to false to skip verifying SSL certificate when calling API from https server
 - ``ssl_ca_cert`` - set this to customize the certificate file to verify the peer
+- ``cert_file`` - set this to customize the certificate client certificate file
+- ``key_file`` - set this to customize the certificate client key file
+- ``key_password`` - set this to customize the certificate client key file password
```
Comment (bednar) [inline README.rst:202]: Please clarify docs and rename options:

```suggestion
- ``cert_file`` - path to the certificate that will be used for mTLS authentication
- ``cert_key_file`` - path to the file contains private key for mTLS certificate
- ``cert_key_password`` - string or function which returns password for decrypting the mTLS private key
```

File: README.rst [hunk 3]
```diff
@@ -226,6 +229,9 @@ Supported properties are:
 - ``INFLUXDB_V2_TIMEOUT`` - socket timeout in ms (default value is 10000)
 - ``INFLUXDB_V2_VERIFY_SSL`` - set this to false to skip verifying SSL certificate when calling API from https server
 - ``INFLUXDB_V2_SSL_CA_CERT`` - set this to customize the certificate file to verify the peer
+- ``INFLUXDB_V2_CERT_FILE`` - set this to customize the certificate client certificate file
+- ``INFLUXDB_V2_KEY_FILE`` - set this to customize the certificate client key file
+- ``INFLUXDB_V2_KEY_PASSWORD`` - set this to customize the certificate client key file password
```
Comment (bednar) [inline README.rst:234]: Please clarify docs and rename options:

```suggestion
- ``INFLUXDB_V2_CERT_FILE`` - path to the certificate that will be used for mTLS authentication
- ``INFLUXDB_V2_CERT_KEY_FILE`` - path to the file contains private key for mTLS certificate
- ``INFLUXDB_V2_CERT_KEY_PASSWORD`` - string or function which returns password for decrypting the mTLS private key
```

File: influxdb_client/client/influxdb_client.py [hunk 4]
```diff
@@ -93,6 +93,9 @@ def from_config_file(cls, config_file: str = "config.ini", debug=None, enable_gz
             - timeout,
             - verify_ssl
             - ssl_ca_cert
+            - cert_file
+            - key_file
+            - key_password
```
Comment (bednar) [inline influxdb_client/client/influxdb_client.py:98]: Please rename options:

```suggestion
            - cert_file
            - cert_key_file
            - cert_key_password
```

File: influxdb_client/client/influxdb_client.py [hunk 5]
```diff
@@ -167,6 +170,9 @@ def from_env_properties(cls, debug=None, enable_gzip=False):
             - INFLUXDB_V2_TIMEOUT
             - INFLUXDB_V2_VERIFY_SSL
             - INFLUXDB_V2_SSL_CA_CERT
+            - INFLUXDB_V2_CERT_FILE
+            - INFLUXDB_V2_KEY_FILE
+            - INFLUXDB_V2_KEY_PASSWORD
```
Comment (bednar) [inline influxdb_client/client/influxdb_client.py:175]: Please rename options:
```suggestion
            - INFLUXDB_V2_CERT_FILE
            - INFLUXDB_V2_CERT_KEY_FILE
            - INFLUXDB_V2_CERT_KEY_PASSWORD
```

File: influxdb_client/client/influxdb_client_async.py [hunk 6]
```diff
@@ -183,6 +183,9 @@ def from_env_properties(cls, debug=None, enable_gzip=False):
             - INFLUXDB_V2_TIMEOUT
             - INFLUXDB_V2_VERIFY_SSL
             - INFLUXDB_V2_SSL_CA_CERT
+            - INFLUXDB_V2_CERT_FILE
+            - INFLUXDB_V2_KEY_FILE
+            - INFLUXDB_V2_KEY_PASSWORD
```
Comment (bednar) [inline influxdb_client/client/influxdb_client_async.py:188]: Please rename options:

```suggestion
            - INFLUXDB_V2_CERT_FILE
            - INFLUXDB_V2_CERT_KEY_FILE
            - INFLUXDB_V2_CERT_KEY_PASSWORD
```

File: influxdb_client/client/_base.py [hunk 7]
```diff
@@ -60,6 +60,10 @@ def __init__(self, url, token, debug=None, timeout=10_000, enable_gzip=False, or
         self.conf.enable_gzip = enable_gzip
         self.conf.verify_ssl = kwargs.get('verify_ssl', True)
         self.conf.ssl_ca_cert = kwargs.get('ssl_ca_cert', None)
+        self.conf.cert_file = kwargs.get('cert_file', None)
+        self.conf.key_file = kwargs.get('key_file', None)
+        self.conf.key_password = kwargs.get('key_password', None)
+        self.conf.ssl_context = kwargs.get('ssl_context', None)
```
Comment (bednar) [inline influxdb_client/client/_base.py:66]: Document this options in: `InfluxDBClient.__init__` and `InfluxDBClientAsync.__init__`.

- https://github.com/influxdata/influxdb-client-python/blob/5ba25c5529109d6afdc4eeadc456a4444ccd7dfc/influxdb_client/client/influxdb_client.py#L27
- https://github.com/influxdata/influxdb-client-python/blob/5ba25c5529109d6afdc4eeadc456a4444ccd7dfc/influxdb_client/client/influxdb_client_async.py#L19

**Review by bednar — CHANGES_REQUESTED** [review]
Thanks for your PR 👍 
There are a few requirements that must be be satisfy before we accept the PR:

File: influxdb_client/client/_base.py (hunk 7 above)
Comment (ZPascal) [reply influxdb_client/client/_base.py]: I've documented the corresponding options and renamed all mTLS-related variables transparently and centrally.

File: influxdb_client/client/influxdb_client.py [hunk 8]
```diff
@@ -40,6 +40,12 @@ def __init__(self, url, token: str = None, debug=None, timeout=10_000, enable_gz
         :param org: organization name (used as a default in Query, Write and Delete API)
         :key bool verify_ssl: Set this to false to skip verifying SSL certificate when calling API from https server.
         :key str ssl_ca_cert: Set this to customize the certificate file to verify the peer.
+        :key str cert_file: Path to the certificate that will be used for mTLS authentication.
+        :key str cert_key_file: Path to the file contains private key for mTLS certificate.
+        :key str cert_key_password: String or function which returns password for decrypting the mTLS private key.
+        :key ssl.SSLContext ssl_context: Specify a custom Python ssl Context for the TLS/ mTlS handshake.
+                                         Be aware that only delivered certificate/ key files or a ssl Context are
+                                         possible.
```
Comment (bednar) [inline influxdb_client/client/influxdb_client.py:48]: Just typo:

```suggestion
        :key ssl.SSLContext ssl_context: Specify a custom Python SSL Context for the TLS/ mTLS handshake.
                                         Be aware that only delivered certificate/ key files or a SSL Context are
                                         possible.
```

File: influxdb_client/client/influxdb_client_async.py [hunk 9]
```diff
@@ -32,6 +32,12 @@ def __init__(self, url, token: str = None, org: str = None, debug=None, timeout=
                             supports the Gzip compression.
         :key bool verify_ssl: Set this to false to skip verifying SSL certificate when calling API from https server.
         :key str ssl_ca_cert: Set this to customize the certificate file to verify the peer.
+        :key str cert_file: Path to the certificate that will be used for mTLS authentication.
+        :key str cert_key_file: Path to the file contains private key for mTLS certificate.
+        :key str cert_key_password: String or function which returns password for decrypting the mTLS private key.
+        :key ssl.SSLContext ssl_context: Specify a custom Python ssl Context for the TLS/ mTlS handshake.
+                                         Be aware that only delivered certificate/ key files or a ssl Context are
+                                         possible.
```
Comment (bednar) [inline influxdb_client/client/influxdb_client_async.py:40]: Just typo:

```suggestion
        :key ssl.SSLContext ssl_context: Specify a custom Python SSL Context for the TLS/ mTLS handshake.
                                         Be aware that only delivered certificate/ key files or a SSL Context are
                                         possible.
```

File: influxdb_client/configuration.py [hunk 10]
```diff
@@ -89,10 +89,15 @@ def __init__(self):
         # client certificate file
         self.cert_file = None
         # client key file
-        self.key_file = None
+        self.cert_key_file = None
```
Comment (bednar) [inline influxdb_client/configuration.py:92]: This is breaking change, please add a note to the `CHANGELOG.md`.

**Review by bednar — CHANGES_REQUESTED** [review]
Thanks again for your PR 👍 There are a few requirements that must be be satisfy before we accept the PR:

1. Rebase your sources with `master` branch
2. Add new client's options (`:key str cert_file`, ...) to the `from_config_file` and `from_env_properties` functions. For more info see: #510

and also:

**Comment by ZPascal** [comment]
Hi @bednar, I'll adapt the PR. I'm also thinking about the possibility to add all SSL-related optional parameters like `verify_ssl `or `ssl_ca_cert` to the docstring documentation of the `from_env_properties` and `from_config_file` function. What do think about the documentation feature?

**Comment by bednar** [comment]
> I'll adapt the PR. I'm also thinking about the possibility to add all SSL-related optional parameters like verify_ssl or ssl_ca_cert to the docstring documentation of the from_env_properties and from_config_file function. What do think about the documentation feature?

Yes, it will be useful. We already improve this docs in #510, please rebase your source to check an actual version of the client.

File: influxdb_client/client/influxdb_client_async.py (hunk 9 above)
Comment (ZPascal) [reply influxdb_client/client/influxdb_client_async.py]: done

File: influxdb_client/client/influxdb_client.py (hunk 8 above)
Comment (ZPascal) [reply influxdb_client/client/influxdb_client.py]: done

File: influxdb_client/client/influxdb_client.py [hunk 11]
```diff
@@ -89,6 +95,14 @@ def from_config_file(cls, config_file: str = "config.ini", debug=None, enable_gz
...
         :key urllib3.util.retry.Retry retries: Set the default retry strategy that is used for all HTTP requests
                                                except batching writes. As a default there is no one retry strategy.
+        :key bool verify_ssl: Set this to false to skip verifying SSL certificate when calling API from https server.
+        :key str ssl_ca_cert: Set this to customize the certificate file to verify the peer.
+        :key str cert_file: Path to the certificate that will be used for mTLS authentication.
+        :key str cert_key_file: Path to the file contains private key for mTLS certificate.
+        :key str cert_key_password: String or function which returns password for decrypting the mTLS private key.
+        :key ssl.SSLContext ssl_context: Specify a custom Python SSL Context for the TLS/ mTLS handshake.
+                                         Be aware that only delivered certificate/ key files or an SSL Context are
+                                         possible.
```
Comment (bednar) [inline influxdb_client/client/influxdb_client.py:105]: The `verify_ssl`, `ssl_ca_cert`, `cert_file`, `cert_key_file`, `cert_key_password` are already presents in **Configuration options**, we can skip it:

```suggestion
        :key ssl.SSLContext ssl_context: Specify a custom Python SSL Context for the TLS/ mTLS handshake.
                                         Be aware that only delivered certificate/ key files or an SSL Context are
                                         possible.
```

File: influxdb_client/client/influxdb_client.py [hunk 12]
```diff
@@ -177,6 +194,14 @@ def from_env_properties(cls, debug=None, enable_gzip=False, **kwargs):
...
```
Comment (bednar) [inline influxdb_client/client/influxdb_client.py:204]: The `verify_ssl`, `ssl_ca_cert`, `cert_file`, `cert_key_file`, `cert_key_password` are already presents in **Configuration options**, we can skip it:

```suggestion
        :key ssl.SSLContext ssl_context: Specify a custom Python SSL Context for the TLS/ mTLS handshake.
                                         Be aware that only delivered certificate/ key files or an SSL Context are
                                         possible.
```

File: influxdb_client/client/influxdb_client_async.py [hunk 13]
```diff
@@ -105,6 +111,14 @@ def from_config_file(cls, config_file: str = "config.ini", debug=None, enable_gz
...
         :key urllib3.util.retry.Retry retries: Set the default retry strategy that is used for all HTTP requests
                                                except batching writes. As a default there is no one retry strategy.
+        :key bool verify_ssl: Set this to false to skip verifying SSL certificate when calling API from https server.
+        :key str ssl_ca_cert: Set this to customize the certificate file to verify the peer.
+        :key str cert_file: Path to the certificate that will be used for mTLS authentication.
+        :key str cert_key_file: Path to the file contains private key for mTLS certificate.
+        :key str cert_key_password: String or function which returns password for decrypting the mTLS private key.
+        :key ssl.SSLContext ssl_context: Specify a custom Python SSL Context for the TLS/ mTLS handshake.
+                                         Be aware that only delivered certificate/ key files or an SSL Context are
+                                         possible.
```
Comment (bednar) [inline influxdb_client/client/influxdb_client_async.py:121]: The `verify_ssl`, `ssl_ca_cert`, `cert_file`, `cert_key_file`, `cert_key_password` are already presents in **Configuration options**, we can skip it:

```suggestion
        :key ssl.SSLContext ssl_context: Specify a custom Python SSL Context for the TLS/ mTLS handshake.
                                         Be aware that only delivered certificate/ key files or an SSL Context are
                                         possible.
```

**Review by bednar — CHANGES_REQUESTED** [review]
We are almost ready to merge this PR, but there are a few requirements that must be be satisfy before we accept the PR:

File: influxdb_client/client/influxdb_client.py (hunk 11 above)
Comment (ZPascal) [reply influxdb_client/client/influxdb_client.py]: done

File: influxdb_client/client/influxdb_client.py (hunk 12 above)
Comment (ZPascal) [reply influxdb_client/client/influxdb_client.py]: done

File: influxdb_client/client/influxdb_client_async.py (hunk 13 above)
Comment (ZPascal) [reply influxdb_client/client/influxdb_client_async.py]: done

File: influxdb_client/client/influxdb_client_async.py [hunk 14]
```diff
@@ -193,6 +210,15 @@ def from_env_properties(cls, debug=None, enable_gzip=False, **kwargs):
...
```
[the remark this answers is not in the document]
Comment (ZPascal) [reply influxdb_client/client/influxdb_client_async.py]: done

**Review by bednar — APPROVED** [review]
LGTM 🚀

**Pull request merged**